File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int size = 0;
        ListNode tmp = head;

        while(tmp != null) {
            size++;
            tmp = tmp.next;
        }

        int removeIndex = size - n;
        if(removeIndex == 0) return head.next;

        tmp = head;
        for(int i = 0; i < size; i++) {
            if(i+1 == removeIndex) {
                tmp.next = tmp.next.next;
                break;
            }
            tmp = tmp.next;
        }
        return head;
    }
}

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-2.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        copyToNew = collections.defaultdict(lambda: Node(0))
        curr = head
        copyToNew[None] = None # prevents the end of the list to use the lambda function

        while curr:
            copyToNew[curr].val = curr.val
            copyToNew[curr].next = copyToNew[curr.next]
            copyToNew[curr].random = copyToNew[curr.random]
            curr = curr.next

        return copyToNew[head]